### DISCUSSION TITLE
thread pool exhaustion with NodeJs

### DISCUSSION THREAD

**Discussion by bataras** [libuv/libuv#4148] [Q&A]
Hi all. Is there a way to get visibility into thread pool usage/heuristics when running Node apps? I'm investigating what the effect is on Node if the libuv thread pool is exhuasted with the default UV_THREADPOOL_SIZE=4

It'd be great if I could get periodic log output in production without having to modify code. But I'm up to the task of modifying code if necessary.

Much appreciated

**Comment by bnoordhuis** [comment]
The way you formulate it doesn't make sense on the face of it. The thread pool doesn't get exhausted, worst case work a work backlog builds up. What specifically do you want to measure?

**Reply by bataras** [reply]
Sorry. I've since looked at the code. Yeah, is there a way to get visibility into the work backlog... the depth of the queue linked list and age of work items as they're completed.

**Reply by bnoordhuis** [reply]
Libuv doesn't record such metrics (and no plans either, in case you're wondering.)

Libuv users like node can record thread pool round-trip time themselves. Unsure if node currently does though.

**Reply by bataras** [reply]
> Libuv users like node record thread pool round-trip time themselves.

Presumably by sending tracer requests like writing to /dev/null

**Reply by bnoordhuis** [reply]
That would definitely work for node users but what I mean is instrumenting uv_fs_t + uv_work_t + uv_random_t requests with metrics inside node itself. Basically this:
```c++
struct FsReqBase {
  // ...
  uv_fs_t req;
  uint64_t start_time;
  hdr_histogram* histogram;
};

void FsReqBase::AfterWork(uv_fs_t* req) {
  FsReqBase* b = FsReqBase::from_req(req);
  uint64_t ns = uv_hrtime() - b->start_time;
  hdr_record_value(b->histogram, ns);
  // ...
}
```
I remember discussing it (I'm also a node maintainer) but I don't think it ever happened.

**Reply by bataras** [reply]
Thanks! I'll play around with it if I get the time. But I think it'd be really good if libuv could formally expose work queue lag and depth. Monitoring tools like Datadog and Prometheus would definitely make it available to production users. As as someone who works in a production, high load environment, I'd love it if it was just "there"

**Reply by bnoordhuis** [reply]
There's probably quite a bit of work involved in exposing such metrics. DD's tracer is pure JS so in order to be able to consume them:

1. Libuv must implement the logic and add a new C API

2. Node must update libuv and add a new JS API

3. And lastly, DD's tracer needs to start using the new API

You're looking at at least one or two days of work. If you or DD wants to sponsor that, get in touch.

**Reply by bataras** [reply]
I might do the PR's myself if I can get the time. that's the hard part.
